blockchain/chain_sync/src/sync.rs

```diff
@@ -0,0 +1,497 @@
+        self.persist_headers(&headers)?;
+
+        Ok(())
+    }
+
+    /// informs the syncer about a new potential tipset
+    /// This should be called when connecting to new peers, and additionally
+    /// when receiving new blocks from the network
+    pub fn inform_new_head(&self, from: &PeerId, fts: &FullTipset) -> Result<(), Error> {
+        // check if full block is nil and if so return error
+        if fts.blocks().is_empty() {
+            return Err(Error::NoBlocks);
+        }
+        // validate message data
+        for block in fts.blocks() {
+            self.validate_msg_data(block)?;
+        }
+
+        // compare target_weight to heaviest weight stored; ignore otherwise
+        let heaviest_tipset = self.chain_store.heaviest_tipset();
+        let best_weight = heaviest_tipset.blocks()[0].weight();
+        let target_weight = fts.blocks()[0].header().weight();
+
+        if !target_weight.lt(&best_weight) {
+            // Store incoming block header
+            self.chain_store.persist_headers(&fts.tipset()?)?;
+            // Set peer head
+            self.sync_manager.set_peer_head(from, fts.tipset()?);
+        }
+        // incoming tipset from miners does not appear to be better than our best chain, ignoring for now
+        Ok(())
+    }
+    /// Validates message root from header matches message root generated from the
+    /// bls and secp messages contained in the passed in block and stores them in a key-value store
+    fn validate_msg_data(&self, block: &Block) -> Result<(), Error> {
+        let sm_root = self.compute_msg_data(block)?;
+        if block.header().messages() != &sm_root {
+            return Err(Error::InvalidRoots);
+        }
+
+        self.chain_store.put_messages(block.bls_msgs())?;
+        self.chain_store.put_messages(block.secp_msgs())?;
+
+        Ok(())
+    }
+    /// Returns message root CID from bls and secp message contained in the param Block
+    fn compute_msg_data(&self, block: &Block) -> Result<Cid, Error> {
+        // collect bls and secp cids
+        let bls_cids = cids_from_messages(block.bls_msgs())?;
+        let secp_cids = cids_from_messages(block.secp_msgs())?;
+        // generate AMT and batch set message values
+        let bls_root = AMT::new_from_slice(self.chain_store.blockstore(), &bls_cids)?;
+        let secp_root = AMT::new_from_slice(self.chain_store.blockstore(), &secp_cids)?;
+
+        let meta = TxMeta {
+            bls_message_root: bls_root,
+            secp_message_root: secp_root,
+        };
+        // store message roots and receive meta_root
+        let meta_root = self.chain_store.blockstore().put(&meta)?;
+
+        Ok(meta_root)
+    }
+    /// Returns FullTipset from store if TipSetKeys exist in key-value store otherwise requests FullTipset
+    /// from block sync
+    pub fn fetch_tipset(&self, _peer_id: PeerId, tsk: &TipSetKeys) -> Result<FullTipset, Error> {
+        let fts = match self.load_fts(tsk) {
```

Comment:
cant you just use `?` instead of match? Just a nit dun matter

Reply:
not changed in this PR but the idea is leaving like this until the TODO is done (I commented the same thing on Dustin's PR)